src/data/buffer/parser/basic.lean

```diff
@@ -0,0 +1,541 @@
+/-
+Copyright (c) 2020 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import data.list.basic
+
+/-!
+# Parsers
+
+`parser α` is the type that describes a computation that can ingest a `char_buffer`
+and output, if successful, a term of type `α`.
+This file expands on the definitions in the core library, proving that all the core library
+parsers are `valid`. There are also lemmas on the composability of parsers.
+
+## Main definitions
+
+* 'parse_result.pos` : The position of a `char_buffer` at which a `parser α` has finished.
+* `parser.valid` : The property that a parser only moves forward within a buffer,
+  in both cases of success or failure.
+
+## Implementation details
+
+Lemmas about how parsers are valid are in the `valid` namespace. That allows using projection
+notation for shorter term proofs that are parallel to the definitions of the parsers in structure.
+
+-/
+
+open parser parse_result
+
+/--
+For some `parse_result α`, give the position at which the result was provided, in either the
+`done` or the `fail` case.
+-/
+@[simp] def parse_result.pos {α} : parse_result α → ℕ
+| (done n _) := n
+| (fail n _) := n
+
+namespace parser
+
+section defn_lemmas
+
+variables {α β : Type} (msgs : thunk (list string)) (msg : thunk string)
+variables (p q : parser α) (cb : char_buffer) (n n' : ℕ) {err : dlist string}
+variables {a : α} {b : β}
+
+/--
+A `parser α` is defined to be `valid` if the result `p cb n` it gives,
+for some `cb : char_buffer` and `n : ℕ`, (whether `done` or `fail`),
+is always at a `parse_result.pos` that is at least `n`. Additionally, if the position `n` provided
+was within the size of the `cb`, then the `parse_result.pos` of the output will also be within
+the size of the `cb`.
+-/
+def valid : Prop :=
+  ∀ (cb : char_buffer) (n : ℕ), n ≤ (p cb n).pos ∧ ((p cb n).pos ≤ cb.size → n ≤ cb.size)
+
+lemma fail_iff :
+  (∀ pos' result, p cb n ≠ done pos' result) ↔
+    ∃ (pos' : ℕ) (err : dlist string), p cb n = fail pos' err :=
+begin
+  cases p cb n,
+  { simp only [not_not, exists_false, exists_eq_right', iff_false, exists_eq', not_forall] },
+  { simp only [forall_const, ne.def, not_false_iff, exists_eq_right', exists_eq', forall_true_iff] }
+end
+
+lemma success_iff :
+  (∀ pos' err, p cb n ≠ fail pos' err) ↔ ∃ (pos' : ℕ) (result : α), p cb n = done pos' result :=
+begin
+  cases p cb n,
+  { simp only [exists_eq_right', exists_eq', forall_const, ne.def, not_false_iff,
+               forall_true_iff] },
+  { simp only [exists_eq_right', exists_eq', exists_prop, not_not, exists_false, iff_false,
+               not_forall] }
+end
+
+variables {p q cb n n' msgs msg}
+
+lemma decorate_errors_fail (h : p cb n = fail n' err) :
+  @decorate_errors α msgs p cb n = fail n ((dlist.lazy_of_list (msgs ()))) :=
+begin
+  dunfold decorate_errors,
+  rw h,
+  refl
+end
```

Comment:
The many uses of `dunfold decorate_errors` worry me a little and seem to lead to funny/brittle proofs. Every time you do this you expose the internal `decorate_errors._match_1`.
What about 
```suggestion
simp [decorate_errors, h]
```
?